### PR TITLE
PoC for T253248

### PR DESCRIPTION
**Pull request by artbaltai**
This is not the final version, but it works.

Actual patch for MediaWiki Core: https://gerrit.wikimedia.org/r/c/mediawiki/core/+/601650

### CODE REVIEW

File: src/Installer.php [hunk 1]
```diff
@@ -28,6 +28,13 @@
 class Installer extends MysqlInstaller {
 	public $supportedEngines = [ 'InnoDB' ];
 
+	/**
+	 * @return string
+	 */
+	public function getReadableName() {
+		return 'Percoda DB';
```
Comment (cicalese) [inline src/Installer.php:35]: typo: Percoda -> Percona

File: i18n/en.json [hunk 2]
```diff
@@ -8,6 +8,7 @@
 	"config-header-percona": "Percona XtraDB Cluster settings",
 	"config-percona-engine": "The only choice for now is InnoDB",
 	"config-percona-innodb": "This is your only choice",
-	"config-percona-engine-help": "What do you need help with?"
+	"config-percona-engine-help": "What do you need help with?",
+	"config-header-PerconaDB": "Percona DB extension"
```
Comment (cicalese) [inline i18n/en.json:12]: where is this used?

**Comment by hexmode** [comment]
Tested against MW master and found the following:

* LocalSettings.php was generated with `$wgDBtype` set to PerconaDB when it should have been Percona.
* LocalSettings.php did not have `wfLoadExtension( "PerconaDB" );` added.

Fixing the above two issues would probably be enough for 1.35.

* There should be a way to specify a class to use for the DB driver.  This would address the first issue and allow better use of namespaces.
* The message keys have mixed case.  I think that is wrong, but we'd have to ask Nikerabbit or someone else in the i18n world to be sure.

With your work, the manual fixes in the first two points above, I was able to merge your work and add a couple more fixes. See 416565c59d7b049db9aa83511bb2c85e8579baff for my changes that I needed.

File: i18n/en.json (hunk 2 above)
Comment (artbaltai) [reply i18n/en.json:12]: Database option form header on the installer page
http://enwiki.web.mw.localhost/mediawiki/mw-config/index.php?page=DBConnect

image

Comment (cicalese) [reply i18n/en.json:12]: But, that doesn't use the message text ("Percona DB extension"). It uses the mis-spelled "Percoda DB" from the getReadableName() function.

**Comment by artbaltai** [comment]
> * LocalSettings.php was generated with `$wgDBtype` set to PerconaDB when it should have been Percona.

fixed with a new PR (also i18n changes needed)
> * LocalSettings.php did not have `wfLoadExtension( "PerconaDB" );` added.

I cannot repeat this case, but solve other problems

> * There should be a way to specify a class to use for the DB driver.  This would address the first issue and allow better use of namespaces.

Are you talking about
`class_alias( "MediaWiki\\Extension\\PerconaDB\\Database", "DatabasePercona" );`
?
It looks like too deep a change and depends on LocalSettings:
`$wgDBtype = "percona";`
in the core work (outside the installer)

> * The message keys have mixed case.  I think that is wrong, but we'd have to ask Nikerabbit or someone else in the i18n world to be sure.

fixed

----
I tested most cases with the PerconaDB extension:
- update LocalSettings.php+PerconaDB 
- update without LocalSettings.php, but with an existing DB tables
- update without LocalSettings.php and clear the database
It seems to have worked

I will now work on the reengineering of this part
https://gerrit.wikimedia.org/r/c/mediawiki/core/+/601650/13/includes/installer/InstallerDbSupport.php#238
and not sure how (this may take some time)

**Comment by hexmode** [comment]
Art <notifications@github.com> writes:

>> * There should be a way to specify a class to use for the DB driver.  This would address the first issue and allow better use of namespaces.
> Are you talking about
> `class_alias( "MediaWiki\\Extension\\PerconaDB\\Database", "DatabasePercona" );`
> ?
> It looks like too deep a change and depends on LocalSettings:
> `$wgDBtype = "percona";`
> in the core work (outside the installer)

It think this could be solved by specifying (in the extension.json) what
$wgDBtype the extension provides and then pointing to the appropriate
class.

For instance, extension.json right now has:

	"Installer": {
		"DbTypes": {
			"percona": {
				"DbInstaller": "MediaWiki\\Extension\\PerconaDB\\Installer"
			}
		}
	},

We could change this to

	"Provides": {
		"DbTypes": {
			"percona": {
				"DbInstaller": "MediaWiki\\Extension\\PerconaDB\\Installer"
				"DbDriver": "MediaWiki\\Extension\\PerconaDB\\Database"
			}
		}
	},


Thanks again!

**Comment by killev** [comment]
@hexmode,

At the moment, the patch provides the capability to implement database extensions.

I created a task on your request and we'll sort it out on regular basis.
 
https://phabricator.wikimedia.org/T255151

**Pull request merged**